src/front/wgsl/mod.rs

```diff
@@ -876,25 +877,38 @@ impl Parser {
                     }
                     lexer.expect(Token::Paren(')'))?;
                     let expr = if components.is_empty() {
-                        let last_component_inner = ctx.resolve_type(last_component)?;
-                        match (&inner, last_component_inner) {
-                            (crate::TypeInner::Scalar { .. }, crate::TypeInner::Scalar { .. })
-                            | (crate::TypeInner::Matrix { .. }, crate::TypeInner::Matrix { .. })
-                            | (crate::TypeInner::Vector { .. }, crate::TypeInner::Vector { .. }) => {
-                                crate::Expression::As {
-                                    expr: last_component,
-                                    kind: kind.ok_or(Error::BadTypeCast(word))?,
-                                    convert: true,
-                                }
+                        // The issue here is that both `ctx.types[handle]` and `last_component_inner` borrow ctx (and one is mutable).
+                        // This isn't pretty.
+
+                        let last_component_ty = {
+                            let last_component_inner = ctx.resolve_type(last_component)?;
+                            match last_component_inner {
+                                crate::TypeInner::Scalar { .. } => Some(0),
+                                crate::TypeInner::Matrix { .. } => Some(1),
+                                crate::TypeInner::Vector { .. } => Some(2),
+                                _ => None,
                             }
+                        };
+                        let ty_type = match &ctx.types[handle].inner {
+                            crate::TypeInner::Scalar { .. } => Some(0),
+                            crate::TypeInner::Matrix { .. } => Some(1),
+                            crate::TypeInner::Vector { .. } => Some(2),
+                            _ => None,
+                        };
+
+                        match (ty_type, last_component_ty) {
```

Comment:
we should just do `if ty_type.is_some() && ty_type == last_component_ty` instead of the match here